packages/langium/src/grammar/validation/validator.ts

```diff
@@ -971,11 +972,23 @@ export class LangiumGrammarValidator {
         const dataTypeRule = isDataTypeRule(rule);
         if (!hasDatatypeReturnType && dataTypeRule) {
             accept('error', 'This parser rule does not create an object. Add a primitive return type or an action to the start of the rule to force object instantiation.', { node: rule, property: 'name' });
-        } else if (hasDatatypeReturnType && !dataTypeRule) {
+        } else if (hasDatatypeReturnType && !dataTypeRule && !rule.fragment) { // fragments are validated in the next check below
             accept('error', 'Normal parser rules are not allowed to return a primitive value. Use a datatype rule for that.', { node: rule, property: rule.dataType ? 'dataType' : 'returnType' });
         }
     }
 
+    checkFragmentKeywords(rule: ast.ParserRule, accept: ValidationAcceptor): void {
+        if (rule.fragment) {
+            if (rule.dataType || rule.returnType || rule.inferredType) {
+                accept('error',
+                    "Fragments assign values to the object of the caller, but don't create a new object themselves. Therefore specifying the type of the returned object is not possible.",
```

Comment:
Suggestion: The validation message is too long IMO.
```suggestion
                    "Fragment rules cannot specify a return type.",
```